pyproject.toml

```diff
@@ -38,7 +38,7 @@ requires-python = ">=3.8"
 dependencies = [
     # attrs 19.2.0 adds `eq` option to decorators
     # attrs 20.1.0 adds @frozen
-    "attrs >= 20.1.0",
+    "attrs >= 21.2.0",
```

Comment:
There's not really any reason to pin this to a version from three years ago instead of 23.2, especially since 21.3 *did* actually make a backwards incompatible change wrt converters.

Reply:
It's usually my habit to base constraints on real compatibility requirements to keep the library as widely-usable for others as possible, but as long as I'm at it, if the team would like like me to constraint to >23.2 for other reasons, I'm happy to. Can we reach some consensus on that in this thread?

Reply:
For others: as @Fuyukai [mentioned](https://github.com/python-trio/trio/pull/2945#issuecomment-1925433171), I could update all `import attr` to the newer `import attrs`, as well. This does sound like a good update to me IMO, but it requires `attrs` 21.3.0. IMO, that's still old enough to seem reasonable, so it's a second reason to make it `attrs>=21.3.0` (in addition to the aforementioned converters change). Anyone have a reason for something even more recent than that? Keep in mind that based on the Changelog, 21.3.0 was the last release that introduced any backward-incompatible changes. This makes it seems like a prime candidate to me.

If anyone's wondering, the converters change, while technically backward-incompatible, is actually one that doesn't require any changes in Trio code (obviously, Trio's been tested for two years since then :P).

Reply:
Or all the way to the latest (23.2) is fine by me, as well. I just want to make sure there's some consensus on that.

Reply:
Actually look at the changes more, there are bug-fixes we really should ensure. I'm signing onto the 23.2 constraint.

Reply:
Added a commit including the `attrs>=23.2.0` constraint _and_ `import attr` -> `import attrs` everywhere.

Reply:
I'd rather pin based on what we require rather than what's latest, because by default people will get the latest (with pip) and all we're causing is incompatibility with things that pin older attrs.

Reply:
> and all we're causing is incompatibility with things that pin older attrs.

This is a fake problem.

Reply:
You say that as though tools like Poetry doesn't put an upper bound by default :'(

(e.g. I clicked 2 arbitrary libraries that depend on `attrs` and one only depended on it in test extras (that were thankfully pinned) and the other used `~=` which has an upper bound)

Reply:
I'm happy with whatever y'all decide.

Reply:
After a few more commits for even more places where more concise syntaxes are now enabled, I believe this discussion about the `attrs` version constraint is the only open question.

Reply:
Running tests with the different versions of `attrs`, I discovered that these changes to make Pyright work actually require `attrs=22.2.0`, because that's when support for the `alias` parameter for `attrs.field()` [was introduced](https://www.attrs.org/en/stable/changelog.html#id7) (note: link correct at time of writing but looks unstable).

Would 22.2.0, only two releases back from current and released in December 2022, be an acceptable middle ground for everyone?

@Fuyukai @A5rocks @TeamSpen210

Reply:
Pinning to 22.2 is what I proposed (in my edit for clarity), so yeah!

Reply:
23.2 switched to using the PEP dataclass transform as opposed to the version only supported by Pyright, which makes it pretty obviously the only correct option in a statically typed Python world.

Reply:
Oh I didn't realize. I don't really know whether bumping versions makes sense then (if one of our deps didn't work on 3.12 at the earliest version we support, would we bump its pin?)

But I'm alright with this change cause at least it's got a technical reason (not +1, not -1).

Reply:
I certainly agree that _users_ should want to be updating to 23.2 for that reason among others, but (and I hate to say this because I'd like to see this PR merged) I'm not so sure that means that it's Trio's place to enforce it on them.